src/lm_speed.cpp

```diff
@@ -0,0 +1,502 @@
+// [[Rcpp::depends(RcppArmadillo)]]
+// [[Rcpp::depends(RcppEigen)]]
+// [[Rcpp::plugins(cpp11)]]
+
+// These functions implement
+// Using Heteroscedasticity Consistent Standard Errors in the Linear Regression Model
+// J. Scott Long and Laurie H. Ervin
+#define ARMA_64BIT_WORD 1
+#include <RcppArmadillo.h>
+#include <RcppEigen.h>
+#include <Rcpp.h>
+using namespace Rcpp;
+
+// Much of what follows is modified from RcppEigen Vignette by Douglas Bates and Dirk Eddelbuettel
+// https://cran.r-project.org/web/packages/RcppEigen/vignettes/RcppEigen-Introduction.pdf
+// [[Rcpp::export]]
+Eigen::MatrixXd AtA(const Eigen::MatrixXd& A) {
+  int n(A.cols());
+  return Eigen::MatrixXd(n,n).setZero().selfadjointView<Eigen::Lower>()
+                             .rankUpdate(A.adjoint());
+}
+
+//' @export
+// [[Rcpp::export]]
+Eigen::MatrixXd mult_diag(const Eigen::MatrixXd& x, const Eigen::ArrayXd& d) {
+
+  Eigen::MatrixXd out(x.rows(), x.cols());
+  for (unsigned j = 0; j < x.cols(); ++j) {
+    out.col(j) = x.col(j) * d(j);
+  }
+
+  return out;
+}
+
+//' @export
+// [[Rcpp::export]]
+Eigen::VectorXd extract_vec(const Eigen::VectorXd& full, const Eigen::VectorXi& ind)
+{
+  int num_indices = ind.sum();
+  Eigen::VectorXd target(num_indices);
+  int j = 0;
+  for (int i = 0; i < full.size(); i++)
+  {
+    if (ind(i) == 1) {
+      target(j) = full(i);
+      j++;
+    }
+  }
+  return target;
+}
+
+//' @export
+// [[Rcpp::export]]
+Eigen::MatrixXd extract_mat_rows(const Eigen::MatrixXd& full, const Eigen::VectorXi& ind)
+{
+  int num_indices = ind.sum();
+  Eigen::MatrixXd target(num_indices, full.cols());
+  int j = 0;
+  for (int i = 0; i < full.rows(); i++)
+  {
+    if (ind(i) == 1) {
+      target.row(j) = full.row(i);
+      j++;
+    }
+  }
+  return target;
+}
```

Comment:
These two functions aren't used right now, but they were how I was subsetting in Eigen before.